gadopt/stokes_integrators.py

```diff
@@ -193,6 +211,62 @@ def setup_solver(self):
         self._solver_setup = True
 
     def solve(self):
+        if not self._initialised:
+            self.initialise()
         if not self._solver_setup:
             self.setup_solver()
+
+        self.solution_old.assign(self.solution)  # Need to update old solution for implicit free surface
+
         self.solver.solve()
+
+
+class FreeSurfaceStokesSolver(StokesSolver):
+    name = 'FreeSurfaceStokesSolver'
+
+    def __init__(self, z, T, approximation, free_surface_dt, free_surface_id, bcs=None, mu=1,
+                 quad_degree=6, cartesian=True, solver_parameters=None, closed=True,
+                 rotational=False, J=None, constant_jacobian=False, **kwargs):
+
+        super().__init__(z, T, approximation, bcs=bcs, mu=mu, quad_degree=quad_degree,
+                         cartesian=cartesian, solver_parameters=solver_parameters, closed=closed,
+                         rotational=rotational, J=J, constant_jacobian=constant_jacobian, **kwargs)
+
+        self.free_surface_dt = free_surface_dt
+        self.free_surface_id = free_surface_id
+
+    def initialise(self):
+        # Solution is made up of 3 trial functions so we have a separate initialise()
+        u, p, eta = fd.split(self.solution)
+        u_old, p_old, eta_old = fd.split(self.solution_old)
+
+        # Theta timestepping scheme for free surface variable
+        # Forward Euler (theta = 0), Crank Nicholson (thetat = 0.5), Backward Euler (theta = 1)
+        theta = 0.5
+        eta_theta = (1-theta)*eta_old + theta*eta
+
+        # Add velocity, pressure and buoyancy term to the fields dictionary
+        self.fields['velocity'] = u
+        self.fields['pressure'] = p
+        self.fields['source'] = self.approximation.buoyancy(p, self.T) * self.k
+
+        # Initialise Stokes equations with a free surface
+        self.equations = FreeSurfaceStokesEquations(self.Z, self.Z, quad_degree=self.quad_degree,
```

Comment:
Your `FreeSurfaceStokesEquations` is actually identical to `StokesEquations`. 
So you can simply write 
```
self.equations = StokesEquation(test_space.sub(2), trial_space.sub(2), quad_degree=quad_degree, **kwargs)
self.equations +=  FreeSurfaceEquation(test_space.sub(2), trial_space.sub(2), quad_degree=quad_degree, **kwargs)
```